firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java

```diff
@@ -0,0 +1,198 @@
+   * element is found (including the prefixed length and the JSON string).
+   *
+   * <p>Once a complete element is read, it is dropped from internal buffer.
+   *
+   * <p>Returns either the bundled element, or null if we have reached the end of the stream.
+   */
+  @Nullable
+  private BundleElement readNextElement() throws IOException, JSONException {
+    int length = readLength();
+    if (length == -1) {
+      return null;
+    }
+
+    String json = readJsonString(length);
+    bytesRead += (int) (Math.log10(length) + 1) + length;
+    return BundleElement.fromJson(serializer, json);
+  }
+
+  /**
+   * Reads the length prefix from the beginning of the internal buffer until the first '{'. Returns
+   * the integer-decoded length.
+   *
+   * <p>If it reached the end of the stream, returns -1.
+   */
+  private int readLength() throws IOException {
+    int nextOpenBracket;
+
+    while ((nextOpenBracket = indexOfOpenBracket()) == -1) {
+      if (!pullMoreData()) {
+        break;
+      }
+    }
+
+    // We broke out of the loop because underlying stream is closed, and there happens to be no
+    // more data to process.
+    if (buffer.remaining() == 0) {
+      return -1;
+    }
+
+    // We broke out of the loop because underlying stream is closed, but still cannot find an
+    // open bracket.
+    if (nextOpenBracket == -1) {
+      raiseError("Reached the end of bundle when a length string is expected.");
+    }
+
+    char[] c = new char[nextOpenBracket];
+    buffer.get(c);
+    return Integer.parseInt(new String(c));
+  }
+
+  /** Returns the index of the first open bracket, or -1 if none is found. */
+  private int indexOfOpenBracket() {
+    buffer.mark();
+    try {
+      for (int i = 0; i < buffer.limit(); ++i) {
+        if (buffer.get() == '{') {
+          return i;
+        }
+      }
+      return -1;
+    } finally {
+      buffer.reset();
+    }
+  }
+
+  /**
+   * Reads from a specified position from the internal buffer, for a specified number of bytes,
+   * pulling more data from the underlying stream if needed.
+   *
+   * <p>Returns a string decoded from the read bytes.
+   */
+  private String readJsonString(int length) throws IOException {
+    char[] c = new char[length];
+
+    int read = Math.min(length, buffer.remaining());
+    buffer.get(c, 0, read);
+
+    while (read < length) {
+      if (!pullMoreData()) {
+        raiseError("Reached the end of bundle when more data was expected.");
+      }
+      int toRead = Math.min(length, buffer.remaining());
+      buffer.get(c, read, toRead);
+      read += toRead;
+    }
+    return new String(c);
+  }
+
+  /**
+   * Pulls more data from underlying stream into the internal buffer.
+   *
+   * @return whether more data is available
+   */
+  private boolean pullMoreData() throws IOException {
+    if (buffer.remaining() == 0) {
+      buffer.compact();
+    }
+
+    int read;
+    do {
+      read = dataReader.read(buffer);
```

Comment:
This loop calls `read` at least twice if not at EOF. I think you could fix it by checking `while (read > 0 && buffer.remaining() > 0)`. That is, if it manages to fill the buffer in a single go, it will fall out of the loop without attempting to read into the full buffer.

Reply:
I updated this to only read once. There is an outer loop that already polls so we don't need this here.